src/loader/bgw_message_queue.c

```diff
@@ -121,25 +121,37 @@ static void
 queue_set_reader(MessageQueue *queue)
 {
 	volatile MessageQueue *vq = queue;
+	pid_t		reader_pid;
 
-	if (queue_get_reader(queue) == InvalidPid)
+	SpinLockAcquire(&vq->mutex);
+	if (vq->reader_pid == InvalidPid)
 	{
-		SpinLockAcquire(&vq->mutex);
 		vq->reader_pid = MyProcPid;
-		SpinLockRelease(&vq->mutex);
 	}
-	else if (queue_get_reader(queue) != MyProcPid)
-		ereport(ERROR, (errmsg("only one reader allowed for TimescaleDB background worker message queue")));
+	reader_pid = vq->reader_pid;
+	SpinLockRelease(&vq->mutex);
+	if (reader_pid != MyProcPid)
+		ereport(ERROR,
+				(errmsg("only one reader allowed for TimescaleDB background worker message queue"),
+				 errhint("Current process is %d", reader_pid)));
 }
 
 static void
 queue_reset_reader(MessageQueue *queue)
 {
 	volatile MessageQueue *vq = queue;
+	bool		reset = false;
 
 	SpinLockAcquire(&vq->mutex);
-	vq->reader_pid = InvalidPid;
+	if (vq->reader_pid == MyProcPid)
+	{
+		reset = true;
+		vq->reader_pid = InvalidPid;
+	}
 	SpinLockRelease(&vq->mutex);
+
+	if (!reset)
+		ereport(ERROR, (errmsg("cannot read if not reader for TimescaleDB background worker message queue")));
```

Comment:
You were just moving this but this error message sounds weird

Reply:
perhaps "multiple TimescaleDB background worker launchers have been started when only one is allowed" hint "please report this bug on our github page"

Reply:
@svenklemm if I'm moving it anyway might as well fix it. What do you think of
```C
"tried to reset TimescaleDB background worker message queue we do not own"
```

Reply:
@JLockerman definitely an improvement over current text